### ISSUE TITLE
Batching : download multiple zoomable images at once automatically

### ISSUE THREAD

**Issue by SB2020-eye** [lovasoa/dezoomify-rs#17]
Hi. Great work on these dezoomers! I was hoping to use the batch capabilities suggested by the dezoomify node application (https://github.com/lovasoa/dezoomify/tree/master/node-app). I'm new at this, and I just couldn't get it to work after many hours. Is there a way to run dezoomify-rs in some kind of batch way?

(And I'm trying to pull images from https://digitalcollections.tcd.ie/home/index.php?DRIS_ID=MS58_003v.)

**Comment by lovasoa** [comment]
dezoomify-rs is a commandline application. You can use it within a [for loop](https://ss64.com/nt/for.html) in a [batch script](https://en.wikibooks.org/wiki/Windows_Batch_Scripting)  in Windows  or a [bash script](https://en.wikibooks.org/wiki/Bash_Shell_Scripting) in Linux, MacOS (or windows with [wsl](https://docs.microsoft.com/en-us/windows/wsl/about)).

For instance, in bash, you could create a file called `urls.txt` containing all the urls you want to dezoomify, and then use [xargs](https://en.wikipedia.org/wiki/Xargs) together with dezoomify-rs : 

```sh
xargs -d '\n' -n 1 ./dezoomify-rs < ./urls.txt
```

**Comment by lovasoa** [comment]
However, I'm not sure what you mean by
> the batch capabilities suggested by the dezoomify node application

The node application downloads a single image at a time (just like dezoomify-rs).

**Comment by SB2020-eye** [comment]
Thanks for the suggestions and links. I'll see if I can figure out a for loop for Windows.

I was trying to make this work: https://github.com/lovasoa/dezoomify/wiki/Node-app-for-shell-and-batch-usage. 
![image](https://user-images.githubusercontent.com/56878280/72668933-e59af880-39f9-11ea-9ebf-21ef80ad5314.png)

**Comment by SB2020-eye** [comment]
(Sorry. I see I cited the wrong "node" page in my original comment. I apologize for the confusion.)

**Comment by lovasoa** [comment]
Yes, you can do the same thing with dezoomify-rs (just replacing `node dezoomify-node.js` with `./dezoomify-rs`)

**Comment by lovasoa** [comment]
Here is a full bash script for downloading all the 339 pages from the book of kells you mention above:

```bash
for i in $(seq -f '%03g' 1 339); do
    for side in r v; do
        ./dezoomify-rs --max-width 10000 "https://digitalcollections.tcd.ie/content/14/pages/MS58_$i$side/image.dzi" "page$i$side.jpg";
    done;
done
```

**Comment by SB2020-eye** [comment]
Wow. I had to erase my next question I was writing. Thank you!

**Comment by lovasoa** [comment]
I am currently running it (it's at page 28). I'll post the result here when it's ready.

**Comment by lovasoa** [comment]
- Pages 0 to 100 : http://dl.free.fr/mkRvsCyMP

**Comment by SB2020-eye** [comment]
Wow again! Thank you so much. I also got it to work - so you've helped me increased my capabilities. :) Much appreciated.

Are pdfs like the ones you linked basically exactly the same in resolution as the jpgs I downloaded with the script you wrote? I don't know how to tell the resolution within a pdf myself, so I'm just checking and curious to know (rather than suppose).

**Comment by lovasoa** [comment]
Yes, the PDF embed the JPEG images without alteration. You can check that simply by extracting an image from the PDF.

**Comment by CavalloScuro** [comment]
Dear all, thank you for these comments and replies. I've attempted to run the bash command outlined by the developers in order to activate a batch dezoomify process, but I haven't been able to overcome some initial errors I'm getting. Here's what I have:

1. A .txt file containing URLS to about 800 zoomify tile .xml files, separated by new line characters (\n)
2. The dezoomify-rs application, unpacked in a directory on my desktop
3. In terminal, I change the directory to the directory on my desktop, which contains dezoomify and my .txt file of URLS
4. Then I type xargs -d '\n' -n 1 ./dezoomify-rs < ./urls.txt
5. I get the following error:

```
xargs -d '\n' -n 1 ./dezoomify-rs < ./urls.txt
xargs: illegal option -- d
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements] [-S replsize]]
             [-J replstr] [-L number] [-n number [-x]] [-P maxprocs]
             [-s size] [utility [argument ...]]
```

I am running macOS Big Sur, which defaults to zsh. I tried typing "bash" before the xargs command, but that doesn't work either. What am I doing wrong?

Thank you in advance for your assistance.

- CS

**Comment by lovasoa** [comment]
https://superuser.com/questions/467176/replacement-for-xargs-d-in-osx

**Comment by hannanaslan** [comment]
> Dear all, thank you for these comments and replies. I've attempted to run the bash command outlined by the developers in order to activate a batch dezoomify process, but I haven't been able to overcome some initial errors I'm getting. Here's what I have:
> 
> 1. A .txt file containing URLS to about 800 zoomify tile .xml files, separated by new line characters (\n)
> 2. The dezoomify-rs application, unpacked in a directory on my desktop
> 3. In terminal, I change the directory to the directory on my desktop, which contains dezoomify and my .txt file of URLS
> 4. Then I type xargs -d '\n' -n 1 ./dezoomify-rs < ./urls.txt
> 5. I get the following error:
> 
> ```
> xargs -d '\n' -n 1 ./dezoomify-rs < ./urls.txt
> xargs: illegal option -- d
> usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements] [-S replsize]]
>              [-J replstr] [-L number] [-n number [-x]] [-P maxprocs]
>              [-s size] [utility [argument ...]]
> ```
> 
> I am running macOS Big Sur, which defaults to zsh. I tried typing "bash" before the xargs command, but that doesn't work either. What am I doing wrong?
> 
> Thank you in advance for your assistance.
> 
> * CS

I am having the same problem. can you please help? [https://www.memoiredeshommes.sga.defense.gouv.fr/fr//_depot_mdh/_depot_images/LEVANT/SHDGR__GR_4_H_5/SHDGR__GR_4_H_5__001/SHDGR__GR_4_H_5__001_0001/ImageProperties.xml] I want to download from 1 to 170.

**Comment by lovasoa** [comment]
See my link above. You can either just remove `-d '\n'` from the command (if none of your urls contains whitespaces) or [install gnu xargs with brew](https://formulae.brew.sh/formula/findutils)

**Comment by CavalloScuro** [comment]
Yes, indeed. Thank you lovasoa. I used the following command and it worked brilliantly.

`xargs -n 1 ./dezoomify-rs -l < ./urls.txt`

The -l is the command for downloading the largest possible image available in the .xml zoomify file. I removed the -d '\n' variables which activated the batch commands.

Thank you.

- CS

**Comment by lovasoa** [comment]
Maybe the server is having some troubles serving all the requests you are making in short succession. You can use the     --parallelism, --retries, --retry-delay and  --timeout options to control that, and be more gentle with the server. Type `dezoomify-rs --help` to view the list of available options.

**Comment by CavalloScuro** [comment]
Thank you. I will look into this right now.

**Comment by piccachilly** [comment]
When I do this, I get the following error 

`ERROR Invalid header value: failed to parse header value` multiple times. 

However, my URLs are formatted correctly in .txt, each with a new line. 

`https://storage.googleapis.com/X/img/dzi/img_1.dzi
https://storage.googleapis.com/X/img/dzi/img_2.dzi`

etc.

For some reason, it will correctly process the last URL in the .txt file. 

Any assistance?

Many thanks!

**Comment by st-carr** [comment]
I'm getting the same error that piccachilly mentioned back in 03/2022. It seems like this is a HTTP issue. Wondering what I'm doing wrong to trigger this error for all URLs except for the last one.

**Comment by rigel71** [comment]
Hi Lovasoa
I'm trying to pull images from https://gallica.bnf.fr/ark:/12148/btv1b9065177d/f1.item.r=landrieux%20jean to https://gallica.bnf.fr/ark:/12148/btv1b9065177d/f464.item.r=landrieux%20jean
how can I create the batch file?
thanks

**Comment by rigel71** [comment]
in linux 
urls.txt with contains
https://gallica.bnf.fr/ark:/12148/btv1b9065177d/f1.highres
till f15.highres
it result
./dezoomify-rs: 1: ./dezoomify-rs: Syntax error: "(" unexpected

**Comment by Arekkusu1998** [comment]
Mrs. @rigel71, Idk if You remember me. We met on GitHub firstly in [this issue](https://github.com/lovasoa/dezoomify/issues/288): we shared an interest in dezoomify ASMI's cadastral maps. Kindly, I would like to ask if I can contact You about those same maps: perhaps You also managed to collect the maps I'm missing, before the old High Quality Digital Library was closed. Is Your [mailbox](https://github.com/lovasoa/dezoomify/issues/288#issuecomment-482561942) the same as it was at that time, please?

**Comment by rigel71** [comment]
Hy,
I remember
yes my mailbox is the same rigel456@virgilio.it
see you soon

**Comment by moroccan69** [comment]
Hey @lovasoa! I hope you're doing well. I've got a bit of a task on my hands and could really use your expertise. I'm trying to download the entire oeuvre by artists from Art & Culture using dezoomify in batch mode. For example, I want to download all 347 images by VVG. (https://artsandculture.google.com/entity/vincent-van-gogh/m07_m2?hl=en) I tried using your tips about using `xargs -d '\n' -n 1 ./dezoomify-rs < ./urls.txt`, but I couldn't manage to make it work. Could you lend a hand or provide some guidance on how to go about this with a code? Your assistance would be greatly appreciated!

**Comment by lovasoa** [comment]
Hello! I'd be happy to setup a small consultancy contract for that! If you are interested, let me know on contact@ophir.dev
